File: strings/map_sum_pairs.rb
class MapSum
    def initialize()
        @h = {}
    end


=begin
    :type key: String
    :type val: Integer
    :rtype: Void
=end
    def insert(key, val)
        @h[key] = val
    end


=begin
    :type prefix: String
    :rtype: Integer
=end
    def sum(prefix)
        @h.keys.select{|x| Regexp.new('^'+prefix) === x}.map{|x| @h[x]}.sum
    end
end

# Your MapSum object will be instantiated and called as such:
# obj = MapSum.new()
# obj.insert(key, val)
# param_2 = obj.sum(prefix)

File: strings/map_sum_pairs.py
class MapSum:

    def __init__(self):
        self.Map = defaultdict(int)
        self.visit = set()
        self.key = {}

    def insert(self, key: str, val: int) -> None:
        cur = ""
        for i in key:
            cur += i
            if key in self.key:
                self.Map[cur] -= self.key[key]
            self.Map[cur] += val

        self.visit.add(key)
        self.key[key] = val

    def sum(self, prefix: str) -> int:
        return self.Map[prefix]


# Your MapSum object will be instantiated and called as such:
# obj = MapSum()
# obj.insert(key,val)
# param_2 = obj.sum(prefix)